File: longest palindromic substring.rb
# Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.

# Example 1:

# Input: "babad"
# Output: "bab"
# Note: "aba" is also a valid answer.


# NOT: LONGEST SHARED SUBSTRING BETWEEN s AND s REVERSED 
# IS: telescoping sames while spreading out to check for mirroreds
def longest_palindrome(s="ac")
	answers=[] #array of palindromes with center = char
	
	if s==""
		return ""
	end

	s.each_char.with_index do |char, idx|
		chopme=s*1.0 # "babad"
		answer=""
		while char==chopme[idx+1]
			answer+=chopme[idx+1]
			chopme[idx+1]=""
			
		end

		if answer.length > s.length/2 
			answers.push(char+answer)
			break
		end
		

		answer=answer.split("")
		answer.unshift(char)

		n=0
		while chopme[idx-1-n]==chopme[idx+1+n] and chopme[idx-1-n]!= nil and chopme[idx+1+n]!=nil and idx-1-n>-1 and idx+1+n<chopme.length
			answer.push(chopme[idx+1+n])
			answer.unshift(chopme[idx-1-n])
			n+=1
		end

		answer=answer.join("")
		answers.push(answer)
	end

	return answers[answers.map{|ele| ele.length}.index(answers.map{|ele| ele.length}.max)]
end

p longest_palindrome("nbnnnnnnbn")

			


	# 	#telescope it, NEIGHBOR HAS TO BE ME or a mirrored sequence so keep deleting same things that touch

File: longest palindromic substring.js
5. Longest Palindromic Substring
Medium

5796

484

Add to List

Share
Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.

Example 1:

Input: "babad"
Output: "bab"
Note: "aba" is also a valid answer.
Example 2:

Input: "cbbd"
Output: "bb"
/**
 * @param {string} s
 * @return {string}
 */

// abcxcba
// abccba

//genius someone elses answer
function longestPalindrome(s){
    let left
    let right
    let answer = ""
    for (let i = 0; i < s.length; i++) {
        for (let j of [0, 1]) {
            left = i
            right = i + j
            while (left >= 0 && s[left] === s[right]){
                left--
                right++
            }
            //overshot by 1 on both left and right -1 7
            if (right - left - 1 > answer.length ){
                answer = s.substring(left+1, right)
            }
        }
        if (answer.length > (s.length - i) * 2) break
    }
    return answer
}
    
//someoone elses
// var longestPalindrome = function(s) {
//     let length = s.length;
//     let answer = ""
//     const dp = Array(s.length).fill(null).map(()=> Array(s.length).fill(0))
//     //"aa" i=1 j =1 j =2 i=0 j =0 j=1
//   for (let i = length - 1; i >= 0; i--){
//     for (let j = i; j < length; j++){
//       dp[i][j] = s[i] == s[j] && (j - i < 3 || dp[i + 1][j - 1]);
//             console.log(dp)
//       if (dp[i][j] && (answer == null || j - i + 1 > answer.length)){
//         answer = s.slice(i, j + 1)
//           console.log(answer)
//       }
      
//     }
//   }
//     // console.log(dp)
    
//   return answer
// }
